### ISSUE TITLE
WARNING: py:class reference target not found: typing.Callable[...<snip>

### ISSUE THREAD

**Issue by altendky** [sphinx-doc/sphinx#8015]
**Describe the bug**
```python
@typing.overload
def host(
    func: typing.Callable[..., typing.Awaitable[object]]
) -> typing.Callable[..., object]:
    ...


@typing.overload
def host(
    *, timeout: float = 3
) -> typing.Callable[
    [typing.Callable[..., typing.Awaitable[object]]], typing.Callable[..., object]
]:
    ...


# TODO: not really sure...
# qtrio/_pytest.py:37: error: Overloaded function implementation does not accept all possible arguments of signature 1
# qtrio/_pytest.py:37: error: Overloaded function implementation does not accept all possible arguments of signature 2
@decorator.decorator  # type: ignore
@pytest.mark.usefixtures("qapp", "qtbot")
def host(
    func, timeout=3, *args, **kwargs,
):
```

Results in:

```
/home/altendky/repos/preqtrio/qtrio/_pytest.py:docstring of qtrio.host:: WARNING: py:class reference target not found: typing.Callable[..., typing.Awaitable[object]]
/home/altendky/repos/preqtrio/qtrio/_pytest.py:docstring of qtrio.host:: WARNING: py:class reference target not found: typing.Callable[..., object]
/home/altendky/repos/preqtrio/qtrio/_pytest.py:docstring of qtrio.host:: WARNING: py:class reference target not found: typing.Callable[[typing.Callable[..., typing.Awaitable[object]]], typing.Callable[..., object]]
```

**To Reproduce**
```bash
git clone https://github.com/altendky/qtrio
cd qtrio
git checkout 510815210096284d6dc2716fb40fd507b2b98c73
python3 -m venv venv
venv/bin/pip install --upgrade pip setuptools wheel
venv/bin/pip install -e .[pyside2,docs]
source venv/bin/activate
cd docs
../venv/bin/pip install --upgrade sphinx==3.1.2
rm -rf build; make html
```

**Expected behavior**
The error doesn't happen.

**Your project**
https://github.com/altendky/qtrio

**Screenshots**
![image](https://user-images.githubusercontent.com/543719/88593508-a5ae7100-d02d-11ea-8ba0-bbbd89bf48ec.png)

**Environment info**
- OS: Linux - Ubuntu 20.04
- Python version: 3.8.3
- Sphinx version: 3.1.2
- Sphinx extensions:
  - sphinx.ext.autodoc
  - sphinx.ext.intersphinx
  - sphinx.ext.coverage
  - sphinx.ext.napoleon
  - sphinx_autodoc_typehints
  - sphinx_qt_documentation
  - sphinxcontrib_trio

**Additional context**
https://github.com/altendky/qtrio/pull/105

This is a rework of the code reported to have a similar issue in #8008

**Comment by tk0miya** [comment]
Is this same as #8008?

**Comment by tk0miya** [comment]
Oh, Sorry. I closed this unintendedly. Reopened.

Now I posted a fix #8020 for #8008. I hope it also resolved this. Could you try it please?

**Comment by altendky** [comment]
Sure thing.  They weren't exactly the same report but it does seem likely they are the same problem.  I filed separately due to #8008 seeming...  to be different across the 3.1.2 release...  maybe?  But this one didn't.  But I also ran into some issues reproducing it exactly.

Yes, I just installed #8020 and that got rid of the warnings listed here.  I'll check against #8008 and report back there for those.

Thanks for the work and the quick turnaround.
